File: Java/092.java
import java.util.HashMap;
import java.util.Map;

public class Problem092 {

    private static final Map<Integer, Integer> chains = new HashMap<>();

    public static int squareDigits(int n) {
        int sum = 0;
        while (n > 0) {
            int digit = n % 10;
            sum += digit * digit;
            n /= 10;
        }
        return sum;
    }

    public static int chain(int n) {
        if (chains.containsKey(n)) {
            return chains.get(n);
        }

        int m = n;
        // Use a temporary list to store numbers in the current chain to avoid recomputing
        // and to add them to the 'chains' map once the termination is found.
        java.util.List<Integer> currentChain = new java.util.ArrayList<>();
        currentChain.add(m);

        while (!chains.containsKey(m)) {
            m = squareDigits(m);
            if (!chains.containsKey(m)) { // Avoid adding duplicates if m is already in currentChain
                currentChain.add(m);
            }
        }

        int termination = chains.get(m);
        for (int numInChain : currentChain) {
            chains.put(numInChain, termination);
        }
        return termination;
    }

    public static void main(String[] args) {
        chains.put(1, 1);
        chains.put(89, 89);

        int count89 = 0;
        for (int n = 1; n <= 10_000_000; n++) {
            if (chain(n) == 89) {
                count89++;
            }
        }
        System.out.println(count89);
    }
}

File: Rust/092.rs
use std::collections::HashMap;

fn square_digits(mut n: u32) -> u32 {
    let mut sum = 0;
    while n > 0 {
        let digit = n % 10;
        sum += digit * digit;
        n /= 10;
    }
    sum
}

fn chain(n: u32, chains: &mut HashMap<u32, u32>) -> u32 {
    if let Some(&result) = chains.get(&n) {
        return result;
    }

    let mut current_num = n;
    let mut path = Vec::new();

    while !chains.contains_key(&current_num) {
        path.push(current_num);
        current_num = square_digits(current_num);
    }

    let final_result = *chains.get(&current_num).unwrap();

    for &num_in_path in &path {
        chains.insert(num_in_path, final_result);
    }

    final_result
}

fn main() {
    let mut chains: HashMap<u32, u32> = HashMap::new();
    chains.insert(1, 1);
    chains.insert(89, 89);

    let mut count_89 = 0;
    for n in 1..=10_000_000 {
        if chain(n, &mut chains) == 89 {
            count_89 += 1;
        }
    }
    println!("{}", count_89);
}
